File: frontend/mobile_app/lib/features/catalog/catalog_controller.dart
import 'dart:async';

import 'package:flutter/foundation.dart' hide Category;

import '../../data/catalog_repository.dart';
import '../../models/category.dart';
import '../../models/product.dart';

enum LoadState { idle, loading, ready, error }

/// Owns the private (salesperson-only) catalog browsing state: categories,
/// search query, category filter, and the resulting product list. Browsing
/// here never touches the display — that only happens via the presentation
/// controller's "Show on Screen".
class CatalogController extends ChangeNotifier {
  CatalogController(this._repo) {
    // Keep the catalogue fresh: silently re-hydrate from the backend every 2 min
    // so CMS edits/new products appear without a manual reload.
    _refreshTimer = Timer.periodic(
      const Duration(minutes: 2),
      (_) => refreshSilently(),
    );
  }

  final CatalogRepository _repo;
  Timer? _refreshTimer;

  LoadState state = LoadState.idle;
  List<Category> categories = <Category>[];
  List<Product> products = <Product>[];
  String? selectedCategoryId;
  String query = '';
  String? error;

  /// The last product opened in detail. Lets the router recover the product
  /// screen when go_router drops `state.extra` on a refresh (e.g. the idle
  /// warning), instead of crashing or bouncing to Home.
  Product? lastViewedProduct;

  Future<void> load() async {
    state = LoadState.loading;
    notifyListeners();
    try {
      categories = await _repo.categories();
      products = await _repo.products();
      state = LoadState.ready;
    } catch (e) {
      error = 'Unable to load the collection.';
      state = LoadState.error;
    }
    notifyListeners();
  }

  Future<void> selectCategory(String? id) async {
    selectedCategoryId = id;
    await _refresh();
  }

  Future<void> search(String q) async {
    query = q;
    await _refresh();
  }

  Future<void> _refresh() async {
    state = LoadState.loading;
    notifyListeners();
    try {
      products = await _repo.products(
        categoryId: selectedCategoryId,
        query: query,
      );
      state = LoadState.ready;
    } catch (e) {
      error = 'Unable to load the collection.';
      state = LoadState.error;
    }
    notifyListeners();
  }

  /// Re-fetch the catalogue in the background without flashing a loading state.
  Future<void> refreshSilently() async {
    try {
      final List<Category> cats = await _repo.categories();
      final List<Product> prods = await _repo.products(
        categoryId: selectedCategoryId,
        query: query,
      );
      categories = cats;
      products = prods;
      if (state != LoadState.error) state = LoadState.ready;
      notifyListeners();
    } catch (_) {
      // Keep showing the last good data on a transient failure.
    }
  }

  @override
  void dispose() {
    _refreshTimer?.cancel();
    super.dispose();
  }
}

File: frontend-rn/mobile_app/src/features/catalog/catalogController.ts
import { Listenable } from '../../core/listenable';
import { CatalogRepository } from '../../data/catalogRepository';
import { Category } from '../../models/category';
import { Product } from '../../models/product';

export enum LoadState {
  idle = 'idle',
  loading = 'loading',
  ready = 'ready',
  error = 'error',
}

/// Owns the private (salesperson-only) catalog browsing state. Browsing never
/// touches the display. Ported 1:1 from `CatalogController`.
export class CatalogController extends Listenable {
  state = LoadState.idle;
  categories: Category[] = [];
  products: Product[] = [];
  selectedCategoryId: string | null = null;
  query = '';
  error: string | null = null;
  /// The last product opened in detail (route recovery aid).
  lastViewedProduct: Product | null = null;

  private refreshTimer: ReturnType<typeof setInterval> | null = null;

  constructor(private readonly repo: CatalogRepository) {
    super();
    this.refreshTimer = setInterval(() => void this.refreshSilently(), 2 * 60 * 1000);
  }

  async load(): Promise<void> {
    this.state = LoadState.loading;
    this.notifyListeners();
    try {
      this.categories = await this.repo.categories();
      this.products = await this.repo.products();
      this.state = LoadState.ready;
    } catch {
      this.error = 'Unable to load the collection.';
      this.state = LoadState.error;
    }
    this.notifyListeners();
  }

  async selectCategory(id: string | null): Promise<void> {
    this.selectedCategoryId = id;
    await this.refreshInternal();
  }

  async search(q: string): Promise<void> {
    this.query = q;
    await this.refreshInternal();
  }

  refresh(): Promise<void> {
    return this.refreshInternal();
  }

  private async refreshInternal(): Promise<void> {
    this.state = LoadState.loading;
    this.notifyListeners();
    try {
      this.products = await this.repo.products({
        categoryId: this.selectedCategoryId ?? undefined,
        query: this.query,
      });
      this.state = LoadState.ready;
    } catch {
      this.error = 'Unable to load the collection.';
      this.state = LoadState.error;
    }
    this.notifyListeners();
  }

  async refreshSilently(): Promise<void> {
    try {
      const [cats, prods] = [
        await this.repo.categories(),
        await this.repo.products({ categoryId: this.selectedCategoryId ?? undefined, query: this.query }),
      ];
      this.categories = cats;
      this.products = prods;
      if (this.state !== LoadState.error) this.state = LoadState.ready;
      this.notifyListeners();
    } catch {
      // Keep showing the last good data on a transient failure.
    }
  }

  dispose(): void {
    if (this.refreshTimer) clearInterval(this.refreshTimer);
  }
}
